File: Tests/YToolsTests/PanelCommandRouterTests.swift
import XCTest
import YToolsCore

final class PanelCommandRouterTests: XCTestCase {
    private let router = PanelCommandRouter()

    func testMapsNumberShortcutsToZeroBasedResults() {
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 18, modifiers: .command),
                mode: .launcher
            ),
            .activateResult(0)
        )
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 25, modifiers: .command),
                mode: .launcher
            ),
            .activateResult(8)
        )
    }

    func testKeepsLauncherOnlyCommandsOutOfClipboardMode() {
        XCTAssertNil(
            router.command(
                for: PanelKeyEvent(keyCode: 16, modifiers: .command),
                mode: .clipboard
            )
        )
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 2, modifiers: .command),
                mode: .clipboard
            ),
            .deleteClipboardItem
        )
    }

    func testOptionArrowsTakePriorityOverNavigation() {
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 125, modifiers: .option),
                mode: .launcher
            ),
            .addSelectedToBuffer(moveToNext: true)
        )
    }

    func testMapsSharedNavigationCommands() {
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 36, modifiers: []),
                mode: .launcher
            ),
            .activateSelected
        )
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 36, modifiers: []),
                mode: .clipboard
            ),
            .activateSelected
        )
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 126, modifiers: []),
                mode: .clipboard
            ),
            .moveSelection(-1)
        )
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 53, modifiers: []),
                mode: .launcher
            ),
            .escape
        )
    }

    func testModifiedReturnKeepsItsDedicatedMeaning() {
        XCTAssertEqual(
            router.command(
                for: PanelKeyEvent(keyCode: 36, modifiers: .command),
                mode: .launcher
            ),
            .revealSelected
        )
        XCTAssertNil(
            router.command(
                for: PanelKeyEvent(keyCode: 36, modifiers: .shift),
                mode: .launcher
            )
        )
    }

    func testLeavesStandardTextEditingShortcutsToFieldEditor() {
        for keyCode: UInt16 in [0, 6, 7, 8, 9] { // A, Z, X, C, V
            XCTAssertNil(
                router.command(
                    for: PanelKeyEvent(keyCode: keyCode, modifiers: .command),
                    mode: .launcher
                )
            )
        }
    }
}

File: Tests/YTools.Windows.Tests/PanelCommandRouterTests.cs
using YTools.Core;

namespace YTools.Windows.Tests;

public class PanelCommandRouterTests
{
    private readonly PanelCommandRouter _router = new();

    [Fact]
    public void Enter_ActivatesSelected()
    {
        var command = _router.Command(
            new PanelKeyEvent(0x0D, PanelKeyModifiers.None),
            PanelInputMode.Launcher);

        Assert.Equal(PanelCommandKind.ActivateSelected, command?.Kind);
    }

    [Theory]
    [InlineData(0x30, 0)]
    [InlineData(0x35, 5)]
    [InlineData(0x39, 9)]
    public void CtrlNumber_ActivatesResult(int keyCode, int expectedIndex)
    {
        var command = _router.Command(
            new PanelKeyEvent(keyCode, PanelKeyModifiers.Command),
            PanelInputMode.Launcher);

        Assert.Equal(PanelCommandKind.ActivateResult, command?.Kind);
        Assert.Equal(expectedIndex, command?.Payload);
    }

    [Fact]
    public void AltUp_AddsSelectedToBuffer()
    {
        var command = _router.Command(
            new PanelKeyEvent(0x26, PanelKeyModifiers.Option),
            PanelInputMode.Launcher);

        Assert.Equal(PanelCommandKind.AddSelectedToBuffer, command?.Kind);
    }

    [Fact]
    public void CtrlComma_OpensSettings()
    {
        var command = _router.Command(
            new PanelKeyEvent(0xBC, PanelKeyModifiers.Command),
            PanelInputMode.Clipboard);

        Assert.Equal(PanelCommandKind.OpenSettings, command?.Kind);
    }

    [Fact]
    public void CtrlD_InClipboard_DeletesItem()
    {
        var command = _router.Command(
            new PanelKeyEvent(0x44, PanelKeyModifiers.Command),
            PanelInputMode.Clipboard);

        Assert.Equal(PanelCommandKind.DeleteClipboardItem, command?.Kind);
    }

    [Fact]
    public void CtrlD_InLauncher_ReturnsNull()
    {
        var command = _router.Command(
            new PanelKeyEvent(0x44, PanelKeyModifiers.Command),
            PanelInputMode.Launcher);

        Assert.Null(command);
    }

    [Fact]
    public void Escape_ReturnsEscapeInBothModes()
    {
        Assert.Equal(
            PanelCommandKind.Escape,
            _router.Command(new PanelKeyEvent(0x1B, PanelKeyModifiers.None), PanelInputMode.Launcher)?.Kind);
        Assert.Equal(
            PanelCommandKind.Escape,
            _router.Command(new PanelKeyEvent(0x1B, PanelKeyModifiers.None), PanelInputMode.Clipboard)?.Kind);
    }
}
